### PR TITLE
[bullshark]: do not insert certificates past committed round

### PR DESCRIPTION
**Pull request by andll**
This check ensures that certificates past committed round for the given origin are not inserted in the consensus DAG.

This is believed to be a source of consensus fork we observe in the private testnet:

(1) Primary is assumed to be able to send repeating certificates to ConsensusProtocol::process_certificate
(2) Currently ConsensusProtocol::process_certificate inserts certificate to hashmap unconditionally
(3) When commit is processed, certificates with round *less than* last committed round are removed in ConsensusState::update
(4) When dag is ordered, certificates with round **equal to** last committed round are ignored during sequencing

The current sequence opens loophole when certificate to round **previous to** last committed is submitted into consensus, it might be sequenced again later. This is in line with current logs we observe in testnet.

This PR enforces invariant that dag does not contain nodes with round **less than** last committed for the given node.

### CODE REVIEW

**Comment by mwtian** [comment]
Reposting just to have the discussion in place: @akichidis has a [nice graph](https://github.com/MystenLabs/sui/pull/5058#discussion_r994585528) that illustrates when a new certificate might be received below the last committed round of the same origin. There might be a design decision to be made:

1. If we want to ignore the incoming cert below the last committed round of A, we can make the change as in this PR. This will also make the dag grow in a single direction without its “holes” backfilled, which will simplify range sync (#5058 will likely work as it is).
2. If we do not want to ignore new certificates below last committed for an authority, we will need a more fine grained way to track which certificate has been ordered in the dag, e.g. a map.

**Comment by andll** [comment]
@mwtian this is interesting.

Are you saying that we can have a situation where we want to commit certificate for round R1 for some origin, but then at some point dag might lead us to commit certificate for round R2<R1 for the same origin? If this could be the case, then logic in the `State::update` is also incorrect (since it just deletes certificates), but my hope was that this is not possible

In other words, question is whether graph of certificates can have two certificates for the same author that are not connected to each other by transitive parent-child relation? (My intuition was that voting rules would prevent this, but I did not examine this thorougly)

**Comment by mwtian** [comment]
When an authority misses one or more rounds, the next header it proposes will not have the parent from the same authority. Header parents are required to come from round - 1.

I agree it seems neither approach is taken consistently right now. For dag construction, allowing committing R2<R1 is what the high level algorithm implied in the [Bullshark paper algorithm 2](https://arxiv.org/pdf/2209.05633.pdf). But similar to how we added the restriction that dfs on dag will not go deeper than the GC depth, I wonder if we can add another restriction to disallow committing R2<R1. I don't see this introducing correctness problems, and it may be nice to have the invariant that certificates from the same authority will be committed into the dag in increasing round order.

As for cons, ignoring relatively recent certificates from the dag may degrade the client experience, especially if we want to have clients only submit to one authority eventually. But hopefully this should be rare.

**Comment by asonnino** [comment]
Nice spot!

Here is the problem in pictures:

Screenshot 2022-10-14 at 14 12 54

### Validator A
1. Validator A commits L2. It thus linearises all the green certificates (in a deterministic order) and commits them. It then deletes the yellow certificate.
2. Validator A receives again the yellow certificate and adds it back to the DAG. 
3. Validator A commits L3. It thus linearises all the blue and yellow certificates and commits them.

### Validator B
1. Validator B commits L2. It thus linearises all the green certificates (in a deterministic order) and commits them. It then deletes the yellow certificate.
2. Validator B commits L3. It thus linearises all the blue certificates and commits them. Note that the validator B did not commit the yellow certificate.
3. Validator B receives again the yellow certificate and adds it back to the DAG.

**Review by asonnino — APPROVED** [review]

**Comment by dmitri-perelman** [comment]
Wow, nice pictures! And amazing analysis + the fix. Thank you!
